plutus-ledger-api/src/Plutus/ApiCommon.hs

```diff
@@ -343,3 +347,52 @@ evaluateScriptCounting lv pv verbose ectx p args = swap $ runWriter @LogOutput $
     tell logs
     liftEither $ first CekError $ void res
     pure final
+
+
+{-| A valid parameter name has to be enumeration, bounded, ordered, and
+prettyprintable in a "lowerKebab" way.
+
+Each API version should expose such an enumeration as an ADT and create
+an instance of ParamName out of it.
+-}
+class (Enum a, Ord a, Bounded a) => IsParamName a where
+   showParamName :: a -> String
+   default showParamName :: (Generic a, GIsParamName (Rep a)) => a -> String
+   showParamName = gshowParamName . from
+
+-- | A datatype-generic class to prettyprint 'sums of nullary constructors' in lower-kebab syntax.
+class GIsParamName f where
+    gshowParamName :: f p -> String
+
+instance (GIsParamName a) => GIsParamName (M1 D i a) where
+    gshowParamName (M1 x) = gshowParamName x
+
+instance Constructor i => GIsParamName (M1 C i U1) where
+    gshowParamName = lowerKebab . conName
+      where
+        lowerKebab :: String -> String
+        lowerKebab (h:t) = toLower h : fmap (\c -> if c == '\''
+                                                  then '-'
+                                                  else c
+                                               ) t
+        lowerKebab _ = error "this should not happen"
+
+instance (GIsParamName a, GIsParamName b) => GIsParamName ((:+:) a b) where
+    gshowParamName (L1 x) = gshowParamName x
+    gshowParamName (R1 x) = gshowParamName x
+
+-- | Return all param names given a specific datatype (enumeration).
+enumParamNames :: IsParamName k => [k]
+enumParamNames = [minBound..maxBound]
+
+-- | Tag some costparameter values that arrived from the ledger, as a specific ParamName version.
+tagWithParamNames :: IsParamName k => [Integer] -> [(k, Integer)]
+tagWithParamNames =
+    -- FIXME: since ledger gives always the full list of values, perhaps we should be more
+    -- strict here than the always-succeeding `zip`.
+    Prelude.zip enumParamNames
```

Comment:
Agree with the comment. At least a length check and a call to `error`.

Reply:
Yep, tbh `zip` should be banned. What you want is something like this: https://github.com/input-output-hk/plutus/pull/4330/files#diff-75012e477e97a9af1b161a0e69125f958ec83f6c8c982f971f7bac41b4c6ef8bR204

Which we can put in our prelude. Weirdly there isn't one on hackage except this one: https://hackage.haskell.org/package/safe-0.3.19/docs/Safe-Exact.html#v:zipExactMay, which is also okay.

Reply:
I addressed with a length check